added Harness CI/CD tutorial (#236)
harness/developer-hub: tutorials/deploy-services/harness-cicd-tutorial.md

> # Harness Continuous Integration and Continuous Delivery

Reviewer: Can we change the title e.g the H1 Tag here to `# Build and Deploy a NodJS App to Kubernetes`

Author: Updated the Header Title

### Harness CI/CD Tutorial
We will only be focusing on creating CI/CD pipelines with Harness in this tutorial. We are also a big fan of GitOps, and we have [GitOps-as-a-service](https://harness.io/blog/generally-available-harness-gitops-as-a-service) available at Harness to speed up your deployments. 

Software development teams that adopt a CI/CD pipeline see faster software deployments with more negligible risk and quicker bug identification and remediation. If you are new to CI/CD, it can seem like an overwhelming amount of new terms and processes to learn. But setting up a CI/CD pipeline is easier than you think. 

A CI/CD pipeline helps your team pinpoint where problems in your software are likely to occur – before they reach end users. Additionally, it allows you to release updates more frequently and spend less time testing individual components or modules before releasing them to users. Harness is a leader in CI/CD with impeccable features. Today, we will show you how to implement CI/CD in minutes with a simple setup. 
![CI/CD flowchart](./static/harness-cicd-tutorial/harness_cicd_pipeline.png)

### Pre-Requisites

> - Free [Harness](https://app.harness.io/auth/#/signup/?utm_source=internal&utm_medium=social&utm_campaign=community&utm_content=pavan_cicd_article&utm_term=get-started) account to do CI/CD (on-premise)

Reviewer: Can we change the UTM links to `https://app.harness.io/auth/#/signup/?module=cd&?utm_source=website&utm_medium=harness-developer-hub&utm_campaign=cd-plg&utm_content=get-started`

Author: Changed the link! Thanks for the quick feedback @ravilach

The Dockerfile you see in the repo will be used to build and push our application as an image to the Docker Hub. The next thing is we will build the image and push it to the Docker Hub using the command,

`docker buildx build --platform=linux/arm64 --platform=linux/amd64  -t docker.io/<docker hub username>/<image name>:<tag> --push  -f ./Dockerfile .`

Once the build and push are successful, you can confirm it by going to your Docker Hub account.
![DockerHub image](./static/harness-cicd-tutorial/hello_world_docker_image.jpeg)

You can see the `deployment.yaml` file in the forked repo, which defines the deployment yaml file to help us deploy the application to our Kubernetes cluster. At this point, make sure your Kubernetes cluster is up and running.

Once everything is set, it is time to set up a Harness account to do CI/CD. Create a free Harness account and your first project. Once you sign-up at Harness, you will be presented with the new CI/CD experience and capabilities.

> Add the required connectors, GitHub repo, Docker Hub and secrets, if any. Delegate in Harness is a service/software you need to install/run on the target cluster [Kubernetes cluster in our case] to connect your artifacts, infrastructure, collaboration, verification and other providers with the Harness Manager. When you set up Harness for the first time, you install a Harness Delegate.

Reviewer 2: ```suggestion
4. Add the required connectors, GitHub repo, Docker Hub and secrets, if any. The Harness Delegate is a lightweight worker process that you need to install on the target cluster to connect your artifacts, infrastructure, collaboration, verification, and other providers with the Harness Manager. When you set up Harness for the first time, you install a Harness Delegate.
```

![Harness Connectors](./static/harness-cicd-tutorial/project_connectors.jpeg)

> Select the Continuous Integration module and add the necessary stages and steps, as shown below.

Reviewer 2: ```suggestion
5. Select the Continuous Integration module, and then add the necessary stages and steps, as shown below.
```

> Select the Continuous Integration module and add the necessary stages and steps, as shown below.

Reviewer 2: ```suggestion
5. Select the Continuous Integration module, and then add the necessary stages and steps, as shown below.
```

![cicd with harness images](./static/harness-cicd-tutorial/build_and_test_stage.jpeg)

> ‘Test’ step set-up is done as below,

Reviewer 2: ```suggestion
6. Set up the Test step as shown below.
```

![npm test](./static/harness-cicd-tutorial/configure_run_step.jpeg)

> The “‘Push to Docker Registry” step is as below,

Reviewer 2: ```suggestion
7. Set up the Push to Docker Registry step as shown below.
```

![build and push image](./static/harness-cicd-tutorial/build_push_image.png)

> Next, set up a deployment pipeline.

Reviewer 2: ```suggestion
8. Next, set up a deployment pipeline.
```

![CD overview](./static/harness-cicd-tutorial/deployment_pipeline.jpeg)

> Add the required details in the “Service” tab.

Reviewer 2: ```suggestion
9. Add the required details on the **Service** tab.
```

![deploy service](./static/harness-cicd-tutorial/service_tab.png)

> Define the environment type in the “Environment” tab.

Reviewer 2: ```suggestion
10. Define the environment type in the **Environment** tab.
```

![environment](./static/harness-cicd-tutorial/environment_tab.png)

> Strategise the execution by selecting which deployment you prefer.

Reviewer 2: ```suggestion
11. Strategize the execution by selecting the deployment you prefer.
```

![execution](./static/harness-cicd-tutorial/execution_tab.jpeg)

> Save everything and run the pipeline.

Reviewer 2: ```suggestion
12. Save everything, and then run the pipeline.
```

![run pipeline](./static/harness-cicd-tutorial/run_pipeline.jpeg)

> You can see both CI and CD getting executed one by one with all the steps specified.

Reviewer 2: ```suggestion
You can see both CI and CD being executed one by one with all the steps specified.
```

![CI/CD execution](./static/harness-cicd-tutorial/ci_cd_execution_successful.png)

> **Congratulations!** We successfully built, and tested the application code and deployed it onto our Kubernetes cluster using the Harness platform.

Reviewer 2: ```suggestion
**Congratulations!** You successfully built and tested the application code and deployed it onto the Kubernetes cluster.
```

You can confirm this deployment by using the command `kubectl get pods`
![kubernetes deployments](./static/harness-cicd-tutorial/pods_running.jpeg)

> You can see two replicas running as per our specifications on the `deployment.yaml` file. Also, confirm the same by going to your Kubernetes dashboard. Since I am using Google cloud (GCP), I can see and confirm that there are two pods running.

Reviewer 2: ```suggestion
You can see two replicas running as per the specifications on the `deployment.yaml` file. Also, confirm the same by going to your Kubernetes dashboard. The following image shows two pods running on Google cloud (GCP).
```
